### PR TITLE
[export] fix zero arg export in training_ir

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130990

Fixed TrainingIRToRunDecomp failures for test_tensor_attribute_zero_args and also a few re-tracability failures because run_decomposition does a retracing.

**edit:** also remove the eliminate_dead_code() in _unlift because of one onnx test failure:
a constant tensor attr was lifted as constant_tensor input but it's not used in the graph after aot_autograd due to a short cut in its decomposition. This causes the setattr to be removed by eliminate_dead_code but the graph signature still contains the name of that buffer, which causes an inconsitency between the transformed graph and ep's original signature after _unlift. And it seems that this has happened a few times where some nodes are accidentally removed and we're in an inconsistent state. 

The alternative of removing it would be: every time we call elimiate_dead_code, we verify the consistency of the graph with 1. the graph before transformation and 2. all the meta datas but i think this deserves a complete design.

### CODE REVIEW

File: test/export/test_export.py [hunk 1]
```diff
@@ -5529,7 +5525,6 @@ def forward(self, x):
 
     # original input names aren't retraceable:
     # compilation will succeed, but names won't match forward() signature.
```
Comment (pianpwk) [inline test/export/test_export.py:5527]: very nit: you can remove these comments

File: torch/export/_trace.py [hunk 2]
```diff
@@ -630,10 +630,10 @@ def _compiling_state_context():
     # Overwrite output specs afterwards.
     from torch._dynamo import config as _dynamo_config
     from torch._functorch._aot_autograd.input_output_analysis import _graph_output_names
-    from torch._guards import detect_fake_mode
 
     flat_fake_args = pytree.tree_leaves((fake_args, fake_kwargs))
-    fake_mode = detect_fake_mode(flat_fake_args)
+    fake_mode = torch.export._trace._detect_fake_mode_from_gm(gm)
```
Comment (pianpwk) [inline torch/export/_trace.py:635]: hmm why export._trace?

**Review by pianpwk — APPROVED** [review]
looks great!

File: torch/export/_trace.py (hunk 2 above)
Comment (ydwu4) [reply torch/export/_trace.py:635]: oops, i can change it to utils._detect_fake_mode_from_gm

File: torch/export/exported_program.py [hunk 3]
```diff
@@ -342,13 +346,12 @@ def _decompose_and_get_gm_with_new_signature_constants(
...
-                fake_args.append(node.meta["val"])
-
-        fake_args_unwrapped = pytree.tree_unflatten(fake_args, mod._in_spec)
-        fake_mode = detect_fake_mode(fake_args)
+        fake_args = [
+            node.meta["val"] for node in mod.graph.nodes if node.op == "placeholder"
+        ]
+        fake_mode = torch._export.utils._detect_fake_mode_from_gm(mod)
+        if fake_mode is None:
+            fake_mode = FakeTensorMode(shape_env=ShapeEnv(), export=True)
```
Comment (zhxchen17) [inline torch/export/exported_program.py:354]: @ydwu4 when this is None? I feel if we have at least 1 call_function or at least 1 input node in the graph, this should be always present?

Comment (ydwu4) [reply torch/export/exported_program.py:354]: It's this test:
```
def test_no_tensor_computation_3(self):
    class Module(torch.nn.Module):
        def forward(self, x, y):
            return 5
```
Do we even want to keep this test?

**Pull request closed without merging**